umh-core/pkg/fsm/dataflowcomponent/manager.go

```diff
@@ -0,0 +1,331 @@
+			managerLogger.Debugf("Creating local config for DataFlowComponent %s", cfg.Name)
+			localCfg := DataFlowComponentConfig{
+				Name:          cfg.Name,
+				DesiredState:  cfg.DesiredState,
+				ServiceConfig: cfg.ServiceConfig,
+			}
+
+			managerLogger.Infof("Successfully created DataFlowComponent instance for: %s", cfg.Name)
+			// Create a new DataFlowComponent instance with the BenthosConfigManager interface impl below
+			return NewDataFlowComponent(localCfg, &benthosMgrAdapter{
+				manager:    benthosManager,
+				logger:     managerLogger.Named("BenthosAdapter"),
+				components: make(map[string]DataFlowComponentConfig),
+			}), nil
+		},
+		// Compare DataFlowComponent configs
+		func(instance public_fsm.FSMInstance, cfg config.DataFlowComponentConfig) (bool, error) {
+			dfcInstance, ok := instance.(*DataFlowComponent)
+			if !ok {
+				managerLogger.Errorf("instance is not a DataFlowComponent")
+				return false, fmt.Errorf("instance is not a DataFlowComponent")
+			}
+
+			unchanged := reflect.DeepEqual(dfcInstance.Config, cfg)
+			if unchanged {
+				managerLogger.Debugf("DataFlowComponent %s is unchanged", cfg.Name)
+			} else {
+				managerLogger.Debugf("DataFlowComponent %s has changed and needs to be updated", cfg.Name)
+			}
+			return unchanged, nil
+		},
+		// Set DataFlowComponent config
+		func(instance public_fsm.FSMInstance, cfg config.DataFlowComponentConfig) error {
+			dfcInstance, ok := instance.(*DataFlowComponent)
+			if !ok {
+				managerLogger.Errorf("instance is not a DataFlowComponent")
+				return fmt.Errorf("instance is not a DataFlowComponent")
+			}
+
+			managerLogger.Debugf("Updating DataFlowComponent %s configuration", cfg.Name)
+
+			// Create local DataFlowComponentConfig from config.DataFlowComponentConfig
+			localCfg := DataFlowComponentConfig{
+				Name:          cfg.Name,
+				DesiredState:  cfg.DesiredState,
+				ServiceConfig: cfg.ServiceConfig,
+			}
+			dfcInstance.Config = localCfg
+
+			managerLogger.Debugf("Successfully updated DataFlowComponent %s configuration", cfg.Name)
+			return nil
+		},
+		// Get expected max p95 execution time per instance
+		func(instance public_fsm.FSMInstance) (time.Duration, error) {
+			dfcInstance, ok := instance.(*DataFlowComponent)
+			if !ok {
+				managerLogger.Errorf("instance is not a DataFlowComponent")
+				return 0, fmt.Errorf("instance is not a DataFlowComponent")
+			}
+			duration := dfcInstance.GetExpectedMaxP95ExecutionTimePerInstance()
+			managerLogger.Debugf("Expected max P95 execution time for %s: %v", dfcInstance.Config.Name, duration)
+			return duration, nil
+		},
+	)
+
+	metrics.InitErrorCounter(metrics.ComponentDataFlowCompManager, name)
+	managerLogger.Infof("DataFlowComponentManager created successfully: %s", managerName)
+
+	return &DataFlowComponentManager{
+		BaseFSMManager: baseManager,
+		logger:         managerLogger,
+		benthosManager: benthosManager,
```

Comment:
why does the data flow component manager have a benthos manager? in benthos, this is done differently (each benthos instance has a service, which then has a manager). we should have this similar across all implementations.

there is one exception for benthos with the port-manager, which actually belongs to the manager: this one is not related to any instance though, it is simply a tool for the manager to know its still available ports, as the instance itself cannot determine which port to use.

Reply:
furthermore, we can have now conflicting ports right? multiple DFCs also mean multiple port managers also mean the same ports allocated right?